demo/src/test/TestActivator.java

```diff
@@ -222,6 +225,52 @@ public Integer call() throws Exception {
 			props.setProperty("main.thread", "true");
 			context.registerService(Callable.class.getName(), r, (Dictionary) props);
 			// throws exception ...
+		} else if ("main.thread.callable.two.runs".equals(p)) {
+			final CountDownLatch latch = new CountDownLatch(1);
+			Callable<Integer> r = new Callable<Integer>() {
+
+				@Override
+				public Integer call() throws Exception {
+					try {
+						System.err.println("In main, return 198");
+						return 198;
+					} finally {
+						CompletableFuture.runAsync(() -> {
+							try {
+								Thread.sleep(4000);
+								Callable<Integer> c = new Callable<Integer>() {
+
+									@Override
+									public Integer call() throws Exception {
+										System.err.println("In main, return 500");
+										return 500;
+									}
+
+								};
+								Properties props = new Properties();
+								props.setProperty("main.thread", "true");
+								System.err.println("regsitering first main callable round");
+								ServiceRegistration registerService = context.registerService(Callable.class.getName(),
+									c, (Dictionary) props);
+							} catch (InterruptedException e) {
+								// TODO Auto-generated catch block
+								e.printStackTrace();
+							}
+						});
+					}
+				}
+
+			};
+
+			Properties props = new Properties();
+			props.setProperty("main.thread", "true");
+			System.err.println("regsitering first main callable round");
+			ServiceRegistration registerService = context.registerService(Callable.class.getName(), r,
+				(Dictionary) props);
+			System.err.println("registered first main callable round");
```

Comment:
This is a duplicate output

Reply:
Yes and no. It helps figuring out the order of things in this highly async process.

Reply:
The difference between the two outputs is hard to see "ing" -> "ed".

Reply:
true

Reply:
I rephrased it a bit, so it is easier to differentiate.